### ISSUE TITLE
Rule proposal: `no-optional-call`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#1776]
### Description

I don't think it's useful.

Case 1:

`foo?.bar?.()`, in this case normally you only need to make sure `foo` exists, if `foo` exists, `foo.bar` always exist, so `foo?.bar()` is safe.

Case 2:
`foo.bar?.()`. You are not sure about what `foo` is, so `foo.bar` may not exist, I don't think it's a good idea to use the optional call, should use `typeof foo.bar === 'function'` / `foo instanceof Foo` instead,

Case 3:

`foo?.()` this is even less useful than the cases above, if variable `foo` is undefined, it still throws.

Case 4:

`foo.bar?.() ?? doSomeThingElse()`, this also hiding potential bug since `foo.bar` may return nullish value.

Case 5:

`foo.bar?.() ?? foo.baz?.()`. Normally, these two methods are different, if they are the same `(foo.bar ?? foo.baz)()`, otherwise, `foo.bar ? foo.bar(argumentForBar) : foo.baz(argumentForBaz)`

### Fail

```js
foo?.()
```

```js
foo.bar?.()
```

```js
foo?.bar?.()
```

### Pass

```js
foo?.bar()
```
---
Welcome to post cases that optional calls are prefered.

**Comment by jonahsnider** [comment]
> I don't think it's a good idea to use the optional call, should use `typeof foo.bar === 'function'` / `foo instanceof Foo` instead,

Why?

**Comment by fisker** [comment]
In that example I assume "You are not sure about what foo is", how can you know `foo.bar` is a function.

**Comment by fisker** [comment]
> What if I know bar is a function, but it's optional.

Good point, and this makes `foo.bar?.()` useful too.
